pilot/pkg/security/trustdomain/bundle.go

```diff
@@ -78,11 +83,25 @@ func (t Bundle) ReplaceTrustDomainAliases(principals []string) []string {
 	return principalsIncludingAliases
 }
 
+// replaceTrustDomains replace the given principal's trust domain with the trust domains from the
+// trustDomains list and return the new principals.
 func (t Bundle) replaceTrustDomains(principal string) []string {
+	trustDomainFromPrincipal, err := getTrustDomainOrError(principal)
+	if err != nil {
+		// This should not happen as we already make sure trustDomainFromPrincipal is not * or "" from
+		// the caller. However, if it happens, return the principal as-is.
+		return []string{principal}
+	}
 	principalsForAliases := []string{}
-	principalsForAliases = append(principalsForAliases, replaceTrustDomainInPrincipal(t.TrustDomain, principal))
-	for _, tdAlias := range t.TrustDomainAliases {
-		principalsForAliases = append(principalsForAliases, replaceTrustDomainInPrincipal(tdAlias, principal))
+	for _, td := range t.TrustDomains {
+		// If the trust domain has a prefix * (e.g. *local from *local/ns/foo/ns/bar), keep the principal
+		// as-is for the matched trust domain. For others, replace the trust domain with the new trust domain
+		// or alias.
+		if suffixMatch(td, trustDomainFromPrincipal) {
+			principalsForAliases = append(principalsForAliases, principal)
+		} else {
+			principalsForAliases = append(principalsForAliases, replaceTrustDomainInPrincipal(td, principal))
+		}
```

Comment:
Let's move the logic of appending the original principal outside of the loop. 
* The original principal should always be added. 
* If tdFromPrincipal does not match any TD, simply keep the original principal. 
* If tdFromPrincipal matches any TD, add those that are not included in the original principal. 

We should have corresponding test cases for the above too. 

```
principalsForAliases = append(principalsForAliases, principal)
foundMatch := false
for _, td := range t.TrustDomains {
   if suffixMatch(td, trustDomainFromPrincipal) {
       foundMatch = true
   } 
}
if foundMatch {
   for _, td := range t.TrustDomains {
      if !suffixMatch(td, trustDomainFromPrincipal) {
          principalsForAliases = append(principalsForAliases, replaceTrustDomainInPrincipal(td, principal))
      }
   }
}
```

Reply:
Do we have a separate function for exact match?

Reply:
The code already supports the 3 cases you mentioned and has tests here https://github.com/istio/istio/blob/21ddfc6d4744b0a08dbd2fa628cf3c30ce5904d7/pilot/pkg/security/trustdomain/bundle_test.go#L79 (I should change it to `principals not match any trust domain` instead).

> Do we have a separate function for exact match?

Yes, the code supports exact match, prefix match, and suffix match.